modules/storages/lib/api/v3/file_links/file_link_representer.rb

```diff
@@ -0,0 +1,136 @@
+            href: api_v3_paths.file_link(represented.id)
+          }
+        end
+
+        link :delete, cache_if: -> { user_allowed_to_manage?(represented) } do
+          {
+            href: api_v3_paths.file_link(represented.id),
+            method: :delete
+          }
+        end
+
+        link :creator do
+          {
+            href: api_v3_paths.user(represented.creator_id),
+            title: represented.creator.name
+          }
+        end
+
+        link :originOpen do
+          {
+            href: ::Storages::StorageUrlService.new(represented).call('open').result
+          }
+        end
+
+        link :staticOriginOpen do
+          {
+            href: api_v3_paths.file_link_open(represented.id)
+          }
+        end
+
+        associated_resource :storage
+
+        associated_resource :storageUrl,
+                            skip_render: ->(*) { true },
+                            getter: ->(*) {},
+                            setter: ->(fragment:, **) {
+                              canonical_url = fragment['href'].gsub(/\/+$/, '')
+                              represented.storage = ::Storages::Storage.find_by(host: canonical_url)
+                            }
+
+        associated_resource :container,
+                            v3_path: :work_package,
+                            representer: ::API::V3::WorkPackages::WorkPackageRepresenter
+
+        def _type
+          'FileLink'
+        end
+
+        private
+
+        def user_allowed_to_manage?(model)
+          current_user.allowed_to?(:manage_file_links, model.container.project)
+        end
+
+        def make_origin_data(model)
+          {
+            id: model.origin_id,
+            name: model.origin_name,
+            mimeType: model.origin_mime_type,
+            createdAt: datetime_formatter.format_datetime(model.origin_created_at, allow_nil: true),
+            lastModifiedAt: datetime_formatter.format_datetime(model.origin_updated_at, allow_nil: true),
+            createdByName: model.origin_created_by_name,
+            lastModifiedByName: model.origin_last_modified_by_name
+          }
+        end
+
+        def parse_origin_data(origin_data)
+          {
+            origin_id: origin_data["id"].to_s,
+            origin_name: origin_data["name"],
+            origin_mime_type: origin_data["mimeType"],
+            origin_created_by_name: origin_data["createdByName"],
+            origin_last_modified_by_name: origin_data["lastModifiedByName"],
+            origin_created_at: ::API::V3::Utilities::DateTimeFormatter.parse_datetime(origin_data["createdAt"],
+                                                                                      'originData.createdAt'),
+            origin_updated_at: ::API::V3::Utilities::DateTimeFormatter.parse_datetime(origin_data["lastModifiedAt"],
+                                                                                      'originData.lastModifiedAt')
+          }
+        end
```

Comment:
I would have to try it out to judge whether it is an actual improvement but it might be worthwhile to attempt extracting all this into its own representer.

Reply:
There was already a `parse_date` method. I added a `parse_datetime` in 5fe5fe2390 (or rather I decoupled `parse_date` and `parse_datetime` to be able to use them independently).

Not sure about using an own representer (and I don't know yet how to do it :sweat_smile:). What would be the benefit? @ulferts

Reply:
Jens comment on this:

> I couldn't find a good example for an inline representer that is not used within our helpers built on top of representable. But for the discussed case I'd just try the example provided by representable itself:
> https://trailblazer.to/2.1/docs/representable.html#explicit-representer and for parsing https://trailblazer.to/2.1/docs/representable.html#nested-parsing
>
> Within our stack there is https://github.com/opf/openproject/blob/dev/lib/api/v3/projects/project_representer.rb#L196-L226 which you would need to adapt but as far as I know `#property` will also accept the `getter:` and `setter:`  parameters if it comes to that. But I'd really try something like:
>
> ```
> property :originData, decorator: A::Decoratators::OriginData, class: Hashie::Mash
> ```

Well, I'm giving up on that one. I tried some stuff, but none worked as I wanted. One issue is that the properties are not nested at the same level, making it necessary to define a setter and a getter, and making it cumbersome to read.

Maybe Representer is not the way to go here. [Representer documentation about getter](https://trailblazer.to/2.1/docs/representable.html#representable-function-api-getter) is stating this:
> As helpful as this option is, please do not overuse it. A representer is not a data mapper, but a document transformer. If your underlying data model and your representers diverge too much, consider using a [twin](https://trailblazer.to/2.1/docs/disposable.html) to simplify the representer.

And they later argue that "_Representable is not a data mapping framework_".